From 11d0143b6218a40dbc6ac7fbd62e2789c9de3088 Mon Sep 17 00:00:00 2001
From: "David R. Bild" <david.bild@xaptum.com>
Date: Thu, 25 Feb 2021 23:45:28 -0600
Subject: [PATCH] ip: use netinet, not linux, header field names

---
 src/ip.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/ip.c b/src/ip.c
index a6fcbcc..6da7ed2 100644
--- a/src/ip.c
+++ b/src/ip.c
@@ -262,7 +262,7 @@ enftun_udp6_pull(struct enftun_packet* pkt)
         goto err;
 
     // Have right amount of data for payload
-    if (pkt->size + sizeof(*udph) != ntohs(udph->len))
+    if (pkt->size + sizeof(*udph) != ntohs(udph->uh_ulen))
         goto err;
 
     // Is checksum valid
-- 
2.29.2

